From 8cb4888d8485073d760cb25a6312c812e05b1c27 Mon Sep 17 00:00:00 2001
From: Marc-Antoine Perennou <Marc-Antoine@Perennou.com>
Date: Wed, 31 Aug 2011 14:14:23 +0200
Subject: [PATCH] codegen: Fix generate_struct_equal_function

Or it would return TRUE when it shouldn't for some structs like Gdk.Atom

Signed-off-by: Marc-Antoine Perennou <Marc-Antoine@Perennou.com>
---
 codegen/valaccodebasemodule.vala |   12 ++++--------
 1 files changed, 4 insertions(+), 8 deletions(-)

diff --git a/codegen/valaccodebasemodule.vala b/codegen/valaccodebasemodule.vala
index c9d1417..45e8031 100644
--- a/codegen/valaccodebasemodule.vala
+++ b/codegen/valaccodebasemodule.vala
@@ -2520,16 +2520,12 @@ public abstract class Vala.CCodeBaseModule : CodeGenerator {
 			ccode.close ();
 		}
 
-		if (st.get_fields().size == 0) {
+		if (st.get_fields().size != 0 || st.is_simple_type ()) {
 			// either opaque structure or simple type
-			if (st.is_simple_type ()) {
-				var cexp = new CCodeBinaryExpression (CCodeBinaryOperator.EQUALITY, new CCodeUnaryExpression (CCodeUnaryOperator.POINTER_INDIRECTION, new CCodeIdentifier ("s1")), new CCodeUnaryExpression (CCodeUnaryOperator.POINTER_INDIRECTION, new CCodeIdentifier ("s2")));
-				ccode.add_return (cexp);
-			} else {
-				ccode.add_return (new CCodeConstant ("FALSE"));
-			}
+			var cexp = new CCodeBinaryExpression (CCodeBinaryOperator.EQUALITY, new CCodeUnaryExpression (CCodeUnaryOperator.POINTER_INDIRECTION, new CCodeIdentifier ("s1")), new CCodeUnaryExpression (CCodeUnaryOperator.POINTER_INDIRECTION, new CCodeIdentifier ("s2")));
+			ccode.add_return (cexp);
 		} else {
-			ccode.add_return (new CCodeConstant ("TRUE"));
+			ccode.add_return (new CCodeConstant ("FALSE"));
 		}
 
 		pop_function ();
-- 
1.7.7.rc0.72.g4b5ea.dirty

